File: Final_assembly/fire-alert-system/fire_alert_flutter/lib/screens/alert_screen.dart
import 'package:flutter/material.dart';
import '../services/api_service.dart';
import '../services/websocket_service.dart';
import '../utils/constants.dart';
import '../widgets/alert_popup.dart';

class AlertScreen extends StatefulWidget {
  const AlertScreen({super.key});

  @override
  State<AlertScreen> createState() => _AlertScreenState();
}

class _AlertScreenState extends State<AlertScreen> {
  List<Map<String, dynamic>> _alerts = [];
  bool _loading = true;
  bool _refreshing = false;
  Map<String, dynamic>? _selectedAlert;
  bool _showPopup = false;

  @override
  void initState() {
    super.initState();
    _loadAlerts();
    _setupWebSocket();
  }

  void _setupWebSocket() {
    final wsService = WebSocketService();
    wsService.on('alert', _handleNewAlert);
    wsService.on('alert_resolved', _handleAlertResolved);
  }

  void _handleNewAlert(dynamic data) {
    setState(() {
      if (!_alerts.any((a) => a['id'] == data['id'])) {
        _alerts.insert(0, data);
        _selectedAlert = data;
        _showPopup = true;
      }
    });
  }

  void _handleAlertResolved(dynamic data) {
    setState(() {
      _alerts.removeWhere((a) => a['id'] == data['alertId']);
    });
  }

  Future<void> _loadAlerts() async {
    try {
      final response = await ApiService.getActiveAlerts();
      if (response['success'] == true) {
        setState(() {
          _alerts = (response['data'] as List).cast<Map<String, dynamic>>();
          _loading = false;
          _refreshing = false;
        });
      }
    } catch (error) {
      print('加载预警失败: $error');
      setState(() {
        _loading = false;
        _refreshing = false;
      });
    }
  }

  Future<void> _resolveAlert(String alertId) async {
    try {
      final response = await ApiService.resolveAlert(alertId);
      if (response['success'] == true) {
        setState(() {
          _alerts.removeWhere((a) => a['id'] == alertId);
        });
      }
    } catch (error) {
      print('解决预警失败: $error');
    }
  }

  IconData _getAlertIcon(String type) {
    switch (type) {
      case 'temperature':
        return Icons.device_thermostat;
      case 'smoke':
        return Icons.cloud;
      case 'co':
        return Icons.warning;
      default:
        return Icons.warning;
    }
  }

  Color _getAlertColor(String level) {
    switch (level) {
      case 'warning':
        return const Color(AppConstants.colorWarning);
      case 'alert':
        return const Color(AppConstants.colorDanger);
      case 'critical':
        return const Color(AppConstants.colorCritical);
      default:
        return const Color(AppConstants.colorTextSecondary);
    }
  }

  @override
  Widget build(BuildContext context) {
    if (_loading) {
      return const Scaffold(
        body: Center(child: CircularProgressIndicator()),
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: const Text('预警列表'),
      ),
      body: _alerts.isEmpty
          ? Center(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  Icon(
                    Icons.check_circle,
                    size: 64,
                    color: const Color(AppConstants.colorSuccess),
                  ),
                  const SizedBox(height: 16),
                  const Text(
                    '暂无活跃预警',
                    style: TextStyle(
                      fontSize: 18,
                      color: Color(AppConstants.colorTextSecondary),
                    ),
                  ),
                ],
              ),
            )
          : RefreshIndicator(
              onRefresh: _loadAlerts,
              child: ListView.builder(
                padding: const EdgeInsets.all(12),
                itemCount: _alerts.length,
                itemBuilder: (context, index) {
                  final alert = _alerts[index];
                  final level = alert['level'] as String? ?? 'alert';
                  final type = alert['type'] as String? ?? '';

                  return Card(
                    margin: const EdgeInsets.only(bottom: 12),
                    child: ListTile(
                      leading: Icon(
                        _getAlertIcon(type),
                        color: _getAlertColor(level),
                        size: 32,
                      ),
                      title: Text(
                        alert['location'] ?? '未知位置',
                        style: const TextStyle(fontWeight: FontWeight.bold),
                      ),
                      subtitle: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(alert['message'] ?? '预警消息'),
                          const SizedBox(height: 4),
                          Container(
                            padding: const EdgeInsets.symmetric(
                              horizontal: 8,
                              vertical: 4,
                            ),
                            decoration: BoxDecoration(
                              color: _getAlertColor(level),
                              borderRadius: BorderRadius.circular(4),
                            ),
                            child: Text(
                              level,
                              style: const TextStyle(
                                color: Colors.white,
                                fontSize: 12,
                                fontWeight: FontWeight.w600,
                              ),
                            ),
                          ),
                        ],
                      ),
                      trailing: IconButton(
                        icon: const Icon(Icons.check_circle, color: Color(AppConstants.colorSuccess)),
                        onPressed: () => _resolveAlert(alert['id'] as String),
                      ),
                      onTap: () {
                        setState(() {
                          _selectedAlert = alert;
                          _showPopup = true;
                        });
                      },
                    ),
                  );
                },
              ),
            ),
      floatingActionButton: _showPopup && _selectedAlert != null
          ? AlertPopup(
              alert: _selectedAlert!,
              onClose: () {
                setState(() {
                  _showPopup = false;
                  _selectedAlert = null;
                });
              },
              onResolve: () {
                if (_selectedAlert != null) {
                  _resolveAlert(_selectedAlert!['id'] as String);
                }
                setState(() {
                  _showPopup = false;
                  _selectedAlert = null;
                });
              },
            )
          : null,
    );
  }
}

File: Final_assembly/fire-alert-system/mobile_app/src/screens/AlertScreen.js
/**
 * 预警列表页面
 */

import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  StyleSheet,
  FlatList,
  TouchableOpacity,
  ActivityIndicator,
  RefreshControl
} from 'react-native';
import { useTranslation } from 'react-i18next';
import i18n from '../utils/i18n';
import Icon from 'react-native-vector-icons/MaterialIcons';
import WebSocketService from '../services/websocket';
import { getActiveAlerts, resolveAlert } from '../services/api';
import { COLORS, ALERT_LEVELS } from '../utils/constants';
import AlertPopup from '../components/AlertPopup';

const AlertScreen = () => {
  const { t } = useTranslation();
  const [alerts, setAlerts] = useState([]);
  const [loading, setLoading] = useState(true);
  const [refreshing, setRefreshing] = useState(false);
  const [showPopup, setShowPopup] = useState(false);
  const [selectedAlert, setSelectedAlert] = useState(null);

  useEffect(() => {
    loadAlerts();

    // 订阅WebSocket事件
    WebSocketService.on('alert', handleNewAlert);
    WebSocketService.on('alert_resolved', handleAlertResolved);

    return () => {
      WebSocketService.off('alert', handleNewAlert);
      WebSocketService.off('alert_resolved', handleAlertResolved);
    };
  }, []);

  const loadAlerts = async () => {
    try {
      const response = await getActiveAlerts();
      if (response.success) {
        setAlerts(response.data);
      }
    } catch (error) {
      console.error('加载预警失败:', error);
    } finally {
      setLoading(false);
      setRefreshing(false);
    }
  };

  const handleNewAlert = (alert) => {
    setAlerts(prev => {
      const exists = prev.find(a => a.id === alert.id);
      if (exists) return prev;
      return [alert, ...prev];
    });
    // 显示弹窗
    setSelectedAlert(alert);
    setShowPopup(true);
  };

  const handleAlertResolved = (data) => {
    setAlerts(prev => prev.filter(a => a.id !== data.alertId));
  };

  const handleResolve = async (alertId) => {
    try {
      const response = await resolveAlert(alertId);
      if (response.success) {
        setAlerts(prev => prev.filter(a => a.id !== alertId));
      }
    } catch (error) {
      console.error('解决预警失败:', error);
    }
  };

  const getAlertIcon = (type) => {
    const icons = {
      temperature: 'device-thermostat',
      smoke: 'cloud',
      co: 'warning'
    };
    return icons[type] || 'warning';
  };

  const getAlertColor = (level) => {
    const colors = {
      warning: COLORS.WARNING,
      alert: COLORS.DANGER,
      critical: COLORS.CRITICAL
    };
    return colors[level] || COLORS.TEXT_SECONDARY;
  };

  const renderAlertItem = ({ item }) => (
    <TouchableOpacity
      style={[styles.alertItem, { borderLeftColor: getAlertColor(item.level) }]}
      onPress={() => {
        setSelectedAlert(item);
        setShowPopup(true);
      }}
    >
      <View style={styles.alertHeader}>
        <View style={styles.alertIconContainer}>
          <Icon
            name={getAlertIcon(item.type)}
            size={24}
            color={getAlertColor(item.level)}
          />
        </View>
        <View style={styles.alertContent}>
          <View style={styles.alertTitleRow}>
            <Text style={styles.alertType}>{t(item.type)}</Text>
            <View style={[styles.alertLevelBadge, { backgroundColor: getAlertColor(item.level) }]}>
              <Text style={styles.alertLevelText}>{t(item.level)}</Text>
            </View>
          </View>
          <Text style={styles.alertLocation}>📍 {item.location}</Text>
          <Text style={styles.alertMessage} numberOfLines={2}>
            {typeof item.message === 'object' 
              ? (item.message?.[i18n.language] || item.message?.zh || item.message?.en || t('alertReceived'))
              : (item.message || t('alertReceived'))}
          </Text>
        </View>
        <TouchableOpacity
          style={styles.resolveButton}
          onPress={() => handleResolve(item.id)}
        >
          <Icon name="check-circle" size={28} color={COLORS.SUCCESS} />
        </TouchableOpacity>
      </View>
    </TouchableOpacity>
  );

  if (loading) {
    return (
      <View style={styles.loadingContainer}>
        <ActivityIndicator size="large" color={COLORS.PRIMARY} />
      </View>
    );
  }

  return (
    <View style={styles.container}>
      {alerts.length === 0 ? (
        <View style={styles.emptyContainer}>
          <Icon name="check-circle" size={64} color={COLORS.SUCCESS} />
          <Text style={styles.emptyText}>{t('noActiveAlerts')}</Text>
        </View>
      ) : (
        <FlatList
          data={alerts}
          renderItem={renderAlertItem}
          keyExtractor={(item) => item.id}
          refreshControl={
            <RefreshControl
              refreshing={refreshing}
              onRefresh={loadAlerts}
              colors={[COLORS.PRIMARY]}
            />
          }
          contentContainerStyle={styles.listContent}
        />
      )}

      {showPopup && selectedAlert && (
        <AlertPopup
          alert={selectedAlert}
          visible={showPopup}
          onClose={() => {
            setShowPopup(false);
            setSelectedAlert(null);
          }}
          onResolve={() => {
            handleResolve(selectedAlert.id);
            setShowPopup(false);
            setSelectedAlert(null);
          }}
        />
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.BACKGROUND,
  },
  loadingContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  emptyContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  emptyText: {
    marginTop: 16,
    fontSize: 18,
    color: COLORS.TEXT_SECONDARY,
    textAlign: 'center',
  },
  listContent: {
    padding: 12,
  },
  alertItem: {
    backgroundColor: '#fff',
    borderRadius: 12,
    padding: 16,
    marginBottom: 12,
    borderLeftWidth: 4,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 3,
  },
  alertHeader: {
    flexDirection: 'row',
    alignItems: 'flex-start',
  },
  alertIconContainer: {
    marginRight: 12,
  },
  alertContent: {
    flex: 1,
  },
  alertTitleRow: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 8,
  },
  alertType: {
    fontSize: 18,
    fontWeight: '700',
    color: COLORS.TEXT_PRIMARY,
    marginRight: 8,
  },
  alertLevelBadge: {
    paddingHorizontal: 8,
    paddingVertical: 4,
    borderRadius: 4,
  },
  alertLevelText: {
    color: '#fff',
    fontSize: 12,
    fontWeight: '600',
  },
  alertLocation: {
    fontSize: 16,
    color: COLORS.TEXT_PRIMARY,
    marginBottom: 8,
    fontWeight: '600',
  },
  alertMessage: {
    fontSize: 14,
    color: COLORS.TEXT_SECONDARY,
    lineHeight: 20,
  },
  resolveButton: {
    marginLeft: 8,
  },
});

export default AlertScreen;

